crates/bevy_math/src/primitives/dim3.rs

```diff
@@ -0,0 +1,159 @@
+    pub direction: Direction3d,
+}
+impl Primitive3d for Line3d {}
+
+/// A section of a line along a direction in 3D space.
+#[derive(Clone, Debug)]
+pub struct LineSegment3d {
+    /// The direction of the line
+    pub direction: Direction3d,
+    /// Half the length of the line segment, the segment extends by this amount in both
+    /// the positive and negative direction
+    pub half_length: f32,
+}
+impl Primitive3d for LineSegment3d {}
+
+impl LineSegment3d {
+    /// Get a line segment and translation from a start and end position of a line segment
+    ///
+    /// Panics if start == end
+    pub fn from_start_end(start: Vec3, end: Vec3) -> (Self, Vec3) {
+        let diff = end - start;
+        let length = diff.length();
+        (
+            Self {
+                direction: Direction3d::from_normalized(diff / length),
+                half_length: length / 2.,
+            },
+            (start + end) / 2.,
+        )
+    }
+
+    /// Get the start position of the line
+    pub fn get_start_pos(&self) -> Vec3 {
+        *self.direction * -self.half_length
+    }
+
+    /// Get the end position of the line
+    pub fn get_end_pos(&self) -> Vec3 {
+        *self.direction * self.half_length
+    }
+}
+
+/// A series of connected line segments in 3D space.
+///
+/// For a version without generics: [`BoxedPolyline3d`]
+#[derive(Clone, Debug)]
+pub struct Polyline3d<const N: usize> {
+    /// The vertices of the polyline
+    pub vertices: [Vec3; N],
+}
+impl<const N: usize> Primitive3d for Polyline3d<N> {}
+
+/// A series of connected line segments in 3D space.
+///
+/// For a version without alloc: [`Polyline3d`]
+#[derive(Clone, Debug)]
+pub struct BoxedPolyline3d {
+    /// The vertices of the polyline
+    pub vertices: Box<[Vec3]>,
+}
+impl Primitive3d for BoxedPolyline3d {}
+
+/// A cuboid primitive, more commonly known as a box.
+#[derive(Clone, Copy, Debug)]
+pub struct Cuboid {
+    /// Half of the width, height and depth of the cuboid
+    pub half_extents: Vec3,
+}
+impl Primitive3d for Cuboid {}
+
+impl Cuboid {
+    /// Create a cuboid from the full size of the cuboid
+    pub fn from_size(size: Vec3) -> Self {
+        Self {
+            half_extents: size / 2.,
+        }
+    }
+}
+
+/// A cylinder primitive
+#[derive(Clone, Copy, Debug)]
+pub struct Cylinder {
+    /// The radius of the cylinder
+    pub radius: f32,
+    /// The half height of the cylinder
+    pub half_height: f32,
+}
+impl Primitive3d for Cylinder {}
+
+impl Cylinder {
+    /// Create a cylinder from a radius and full height
+    pub fn from_radius_height(radius: f32, height: f32) -> Self {
+        Self {
+            radius,
+            half_height: height / 2.,
+        }
+    }
+}
+
+/// A capsule primitive
+#[derive(Clone, Copy, Debug)]
+pub struct Capsule {
+    /// The radius of the capsule
+    pub radius: f32,
+    /// Half the height of the capsule, excluding the hemispheres
+    pub half_extent: f32,
```

Comment:
`half_extent` is a bit ambiguous and could also refer to the radius if that didn't already exist. Maybe `half_length`? I've seen some other implementations use it.

What this really is is the length of the segment from A to B where the hemispheres are at those points. Some implementations, [like Parry](https://github.com/dimforge/parry/blob/af38912c942afd260e3d65931c750e57beb7c4b0/src/shape/capsule.rs#L22), store the segment points directly, because I believe capsule collision detection is based on computing the distance from the segment, which requires those points, and it also allows rotated/offset capsules (but that's not required for primitives). Some other implementations [like the one in BepuPhysics](https://github.com/bepu/bepuphysics2/blob/0a11a24d91ad80f4656cb705df010398061863a0/BepuPhysics/Collidables/Capsule.cs#L23C8-L23C8) store just the segment length as well though, so it's probably fine, especially for a general purpose shape.

Reply:
`half_length` would suggest it's half of the length of the capsule, which isn't quite correct either, but I guess `half_extent` is just as confusing. The real definition of a capsule is just a surface at a distance (radius) from a line segment. Since a line segment can be defined in various ways the same applies to capsules. Some representations support different orientations, others don't. In this case it's an upright capsule because anything that cares about orientation will probably already use Transform. I guess I'll explain it in the docs a bit so the user has a better idea of what half_length defines

Reply:
Yeah I guess `half_extent` and `half_length` are similar, but I don't know if you can say "the half extents of the segment", so length might be more fitting. Extent might also confuse people since it's not like the extents of a cuboid or rectangle where it covers the entire shape, but rather specifically the segment length